server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java

```diff
@@ -0,0 +1,225 @@
+        }
+        assert fieldType().isSearchable() == false;
+
+        // SortedMap is expected to be sorted by key (field name)
+        SortedMap<String, BytesReference> dimensionFields = context.doc().getDimensionBytes();
+        if (dimensionFields.isEmpty()) {
+            throw new IllegalArgumentException("Dimension fields are missing.");
+        }
+
+        try (BytesStreamOutput out = new BytesStreamOutput()) {
+            out.writeVInt(dimensionFields.size());
+            for (Map.Entry<String, BytesReference> entry : dimensionFields.entrySet()) {
+                String fieldName = entry.getKey();
+                BytesRef fieldNameBytes = new BytesRef(fieldName);
+                int len = fieldNameBytes.length;
+                if (len > DIMENSION_NAME_LIMIT) {
+                    throw new IllegalArgumentException(
+                        "Dimension name must be less than [" + DIMENSION_NAME_LIMIT + "] bytes but [" + fieldName + "] was [" + len + "]."
+                    );
+                }
+                // Write field name in utf-8 instead of writeString's utf-16-ish thing
+                out.writeBytesRef(fieldNameBytes);
+                entry.getValue().writeTo(out);
+            }
+
+            BytesReference timeSeriesId = out.bytes();
+            if (timeSeriesId.length() > LIMIT) {
+                throw new IllegalArgumentException(NAME + " longer than [" + LIMIT + "] bytes [" + timeSeriesId.length() + "].");
+            }
+            assert timeSeriesId != null : "In time series mode _tsid cannot be null";
+            context.doc().add(new SortedSetDocValuesField(fieldType().name(), timeSeriesId.toBytesRef()));
+        }
+    }
+
+    @Override
+    protected String contentType() {
+        return CONTENT_TYPE;
+    }
+
+    /**
+     * Parse the {@code _tsid} into a human readable map.
+     */
+    public static Map<String, Object> parse(StreamInput in) {
+        try {
+            int size = in.readVInt();
+            Map<String, Object> result = new LinkedHashMap<String, Object>(size);
+
+            for (int i = 0; i < size; i++) {
+                String name = in.readString();
+
+                int type = in.read();
+                switch (type) {
+                    case (byte) 's':
+                        result.put(name, in.readBytesRef().utf8ToString());
+                        break;
+                    case (byte) 'l':
+                        result.put(name, in.readLong());
+                        break;
+                    default:
+                        throw new IllegalArgumentException("Cannot parse [" + name + "]: Unknown type [" + type + "]");
+                }
+            }
+            return result;
+        } catch (IOException | IllegalArgumentException e) {
+            throw new IllegalArgumentException("Error formatting " + NAME + ": " + e.getMessage(), e);
+        }
+    }
+
+    static BytesReference extractTsidValue(String value) throws IOException {
```

Comment:
I'd probably catch `IOException` here and wrap. It's not really likely `BytesStreamOutput` throws at all and it'd make the caller happier.

Reply:
That was my initial implementation. I only moved exception handling one leven up because I could have the field name and throw a nice exception such as `throw new IllegalArgumentException("Dimension field [" + fieldType().name() + "] cannot be serialized.", e);`.

However,  the field name (and its value) will be exposed in the `MapperException` so that I improve the API at the expense of logging